File: cpp/src/telemetry.cpp
#include "hics/telemetry.hpp"

#include <chrono>
#include <cstring>
#include <thread>

namespace hics {

TelemetryRingBuffer::TelemetryRingBuffer() { buffer_.fill({0, 0, 0.0f}); }

void TelemetryRingBuffer::write(uint32_t edge_idx, float utilization) {
    uint32_t idx = write_idx_.fetch_add(1, std::memory_order_relaxed) % TELEMETRY_RING_SIZE;
    buffer_[idx] = {static_cast<uint64_t>(
                        std::chrono::steady_clock::now().time_since_epoch().count()),
                    edge_idx, utilization};
    std::atomic_thread_fence(std::memory_order_release);
}

bool TelemetryRingBuffer::read_latest(uint32_t edge_idx, float& out_util) const {
    uint32_t start = write_idx_.load(std::memory_order_acquire);
    for (uint32_t i = 0; i < TELEMETRY_RING_SIZE; ++i) {
        uint32_t idx = (start + TELEMETRY_RING_SIZE - 1 - i) % TELEMETRY_RING_SIZE;
        if (buffer_[idx].edge_idx == edge_idx) {
            out_util = buffer_[idx].utilization;
            return true;
        }
    }
    return false;
}

std::vector<float> TelemetryRingBuffer::snapshot(size_t num_edges) const {
    std::vector<float> utils(num_edges, 0.0f);
    for (size_t e = 0; e < num_edges; ++e) {
        read_latest(static_cast<uint32_t>(e), utils[e]);
    }
    return utils;
}

TelemetryDaemon::TelemetryDaemon(size_t num_edges)
    : num_edges_(num_edges), simulated_utils_(num_edges, 0.0f) {}

TelemetryDaemon::~TelemetryDaemon() { stop(); }

void TelemetryDaemon::start() {
    if (running_.exchange(true)) return;
    collector_thread_ = std::thread([this] { collect_loop(); });
}

void TelemetryDaemon::stop() {
    if (!running_.exchange(false)) return;
    if (collector_thread_.joinable()) collector_thread_.join();
}

void TelemetryDaemon::set_simulated_util(uint32_t edge_idx, float util) {
    if (edge_idx < simulated_utils_.size()) {
        simulated_utils_[edge_idx] = std::clamp(util, 0.0f, 1.0f);
    }
}

void TelemetryDaemon::collect_loop() {
    using namespace std::chrono;
    const auto interval = microseconds(1000000 / TELEMETRY_SAMPLE_HZ);
    while (running_.load(std::memory_order_relaxed)) {
        for (size_t e = 0; e < num_edges_; ++e) {
            float util = simulated_utils_[e];
            // In production: read_nvlink_util / read_ib_util / read_pcie_util
            ring_.write(static_cast<uint32_t>(e), util);
        }
        std::this_thread::sleep_for(interval);
    }
}

float TelemetryDaemon::read_nvlink_util(uint32_t edge_idx) {
    // Production: nvmlDeviceGetNvLinkUtilizationCounter
    (void)edge_idx;
    return 0.0f;
}

float TelemetryDaemon::read_ib_util(uint32_t edge_idx) {
    // Production: ibverbs perfquery MAD batched every 2 ms
    (void)edge_idx;
    return 0.0f;
}

float TelemetryDaemon::read_pcie_util(uint32_t edge_idx) {
    // Production: pcm-iio PMU counters
    (void)edge_idx;
    return 0.0f;
}

}  // namespace hics

File: rust/src/telemetry.rs
use crossbeam_channel::{bounded, Receiver, Sender};
use parking_lot::RwLock;
use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

pub const TELEMETRY_RING_SIZE: usize = 512;
pub const TELEMETRY_SAMPLE_HZ: u32 = 1000;

#[derive(Clone, Copy, Debug)]
pub struct TelemetrySample {
    pub timestamp_ns: u64,
    pub edge_idx: u32,
    pub utilization: f32,
}

pub struct TelemetryRingBuffer {
    buffer: RwLock<[TelemetrySample; TELEMETRY_RING_SIZE]>,
    write_idx: AtomicU32,
}

impl TelemetryRingBuffer {
    pub fn new() -> Self {
        Self {
            buffer: RwLock::new([TelemetrySample {
                timestamp_ns: 0,
                edge_idx: 0,
                utilization: 0.0,
            }; TELEMETRY_RING_SIZE]),
            write_idx: AtomicU32::new(0),
        }
    }

    pub fn write(&self, edge_idx: u32, utilization: f32) {
        let idx = self.write_idx.fetch_add(1, Ordering::Relaxed) as usize % TELEMETRY_RING_SIZE;
        let ts = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;
        let mut buf = self.buffer.write();
        buf[idx] = TelemetrySample {
            timestamp_ns: ts,
            edge_idx,
            utilization,
        };
    }

    pub fn snapshot(&self, num_edges: usize) -> Vec<f32> {
        let buf = self.buffer.read();
        let start = self.write_idx.load(Ordering::Acquire) as usize;
        let mut utils = vec![0.0f32; num_edges];
        for i in 0..TELEMETRY_RING_SIZE {
            let idx = (start + TELEMETRY_RING_SIZE - 1 - i) % TELEMETRY_RING_SIZE;
            let s = buf[idx];
            if (s.edge_idx as usize) < num_edges {
                if utils[s.edge_idx as usize] == 0.0 {
                    utils[s.edge_idx as usize] = s.utilization;
                }
            }
        }
        utils
    }
}

impl Default for TelemetryRingBuffer {
    fn default() -> Self {
        Self::new()
    }
}

pub struct TelemetryDaemon {
    ring: Arc<TelemetryRingBuffer>,
    running: Arc<AtomicBool>,
    simulated: Arc<RwLock<Vec<f32>>>,
    handle: Option<thread::JoinHandle<()>>,
    cmd_tx: Option<Sender<()>>,
}

impl TelemetryDaemon {
    pub fn new(num_edges: usize) -> Self {
        Self {
            ring: Arc::new(TelemetryRingBuffer::new()),
            running: Arc::new(AtomicBool::new(false)),
            simulated: Arc::new(RwLock::new(vec![0.0; num_edges])),
            handle: None,
            cmd_tx: None,
        }
    }

    pub fn ring(&self) -> Arc<TelemetryRingBuffer> {
        Arc::clone(&self.ring)
    }

    pub fn set_simulated_util(&self, edge_idx: u32, util: f32) {
        let mut sim = self.simulated.write();
        if (edge_idx as usize) < sim.len() {
            sim[edge_idx as usize] = util.clamp(0.0, 1.0);
        }
    }

    pub fn start(&mut self) {
        if self.running.swap(true, Ordering::SeqCst) {
            return;
        }
        let (tx, rx): (Sender<()>, Receiver<()>) = bounded(1);
        self.cmd_tx = Some(tx);

        let ring = Arc::clone(&self.ring);
        let running = Arc::clone(&self.running);
        let simulated = Arc::clone(&self.simulated);

        self.handle = Some(thread::spawn(move || {
            let interval = Duration::from_micros(1_000_000 / TELEMETRY_SAMPLE_HZ as u64);
            while running.load(Ordering::Relaxed) {
                let sim = simulated.read();
                for (e, &util) in sim.iter().enumerate() {
                    ring.write(e as u32, util);
                }
                drop(sim);
                thread::sleep(interval);
                if rx.try_recv().is_ok() {
                    break;
                }
            }
        }));
    }

    pub fn stop(&mut self) {
        self.running.store(false, Ordering::SeqCst);
        if let Some(tx) = self.cmd_tx.take() {
            let _ = tx.send(());
        }
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }
    }
}

impl Drop for TelemetryDaemon {
    fn drop(&mut self) {
        self.stop();
    }
}
